akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala

```diff
@@ -318,55 +318,32 @@ private[akka] final case class Grouped[T](n: Int) extends PushPullStage[T, immut
  * INTERNAL API
  */
 private[akka] final case class Sliding[T](n: Int, step: Int) extends PushPullStage[T, immutable.Seq[T]] {
-  private val buf = {
-    val b = Vector.newBuilder[T]
-    b.sizeHint(n)
-    b
-  }
-  var bufferedElements = 0
+  private var buf = Vector.empty[T]
 
   override def onPush(elem: T, ctx: Context[immutable.Seq[T]]): SyncDirective = {
-    buf += elem
-    bufferedElements += 1
-    if (bufferedElements < n) {
+    buf :+= elem
+    if (buf.size < n) {
+      ctx.pull()
+    } else if (buf.size == n) {
+      ctx.push(buf)
+    } else if (step > n) {
+      if (buf.size == step)
+        buf = Vector.empty
```

Comment:
My thinking was to defer opting it until it becomes a graphstage.
Right now it is buggy and I'd rather have an impl which is easy to reason
about than a buggy fast one. This PR is what my very limited time can offer
right now :-/

-- 
Cheers,
√
On 12 Dec 2015 18:16, "Roland Kuhn" <notifications@github.com> wrote:

> In akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala
> <https://github.com/akka/akka/pull/19153#discussion_r47434165>:
>
> >
> >    override def onPush(elem: T, ctx: Context[immutable.Seq[T]]): SyncDirective = {
> > -    buf += elem
> > -    bufferedElements += 1
> > -    if (bufferedElements < n) {
> > +    buf :+= elem
> > +    if (buf.size < n) {
> > +      ctx.pull()
> > +    } else if (buf.size == n) {
> > +      ctx.push(buf)
> > +    } else if (step > n) {
> > +      if (buf.size == step)
> > +        buf = Vector.empty
>
> This is pretty wasteful for large step values: we would build up a huge
> vector for no reason.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/akka/akka/pull/19153/files#r47434165>.
>